### DISCUSSION TITLE
DOS attack due to automatic Rack body parsing (incorrect Content-Type)

### DISCUSSION THREAD

**Discussion by md-work** [sinatra/sinatra#1916] [Q&A] [answered]
HTTP requests containing a large body can cause huge memory and CPU consumption on the server.
Sending a 4 GB body caused the Ruby process to grow over 40 GB in memory. (then swapping started)

**This can be used by a malicious client to [DOS attack](https://en.wikipedia.org/wiki/Denial-of-service_attack) the server!**

It happens if the `Content-Type` suggests the body can be parsed by [Rack](https://github.com/rack/rack/).
Bad example: `Content-Type: application/x-www-form-urlencoded` 
Good example: `Content-Type: application/octet-stream`
_These are just two example content types which reproduce good and bad behavior._
The content type must not even be wrong. The client might be sending 4 GB of `x-www-form-urlencoded` data.

&nbsp;

## Details

In some way this is a problem of Rack. But Sinatra is invoking Rack to start the parsing. So I'm opening the issue here.  
Stacktrace _(captured via gdb `call (void)rb_backtrace()` on [Debian-12-Testing / Bookworm](https://cdimage.debian.org/cdimage/weekly-live-builds/amd64/iso-hybrid/))_:
```
[...]
/usr/share/rubygems-integration/all/gems/sinatra-3.0.5/lib/sinatra/base.rb:939:in `block in call!'
/usr/share/rubygems-integration/all/gems/sinatra-3.0.5/lib/sinatra/base.rb:1116:in `dispatch!'
/usr/share/rubygems-integration/all/gems/sinatra-3.0.5/lib/sinatra/base.rb:79:in `params'
/usr/share/rubygems-integration/all/gems/rack-2.2.6.4/lib/rack/request.rb:32:in `params'
/usr/share/rubygems-integration/all/gems/rack-2.2.6.4/lib/rack/request.rb:469:in `params'
/usr/share/rubygems-integration/all/gems/rack-2.2.6.4/lib/rack/request.rb:454:in `POST'
[...]
```
In `Rack::Request#POST` this is happening:
```
        elsif form_data? || parseable_data?
          unless set_header(RACK_REQUEST_FORM_HASH, parse_multipart)
            form_vars = get_header(RACK_INPUT).read
```
`parseable_data?` is true depending on the `Content-Type`.

&nbsp;

**Reproduction:**
```bash
truncate --size="$((4*1024**3))" file_4g
wget --tries=1 --post-file=file_4g --debug 'http://127.0.0.1:8080/'`
```
If the client adds ` --header='Content-Type: application/octet-stream'` the problem will be avoided.

&nbsp;

## Questions, Proposal, Related

Is there a way to disable this behavior in Sinatra?
So Sinatra won't ask Rack to parse the body.
When using pure Rack there's no automatic body parsing.

I guess either Sinatra or Rack should check the body size or limit the maximum number of bytes being parsed.
A simple solution would be to add a maximum number of bytes to the `read` call:
```
            form_vars = get_header(RACK_INPUT).read(16*1024**2)  #16 MiB
```

Related: https://github.com/rack/rack/discussions/2049

**Comment by dentarg** [comment]
I will convert this to a discussion because I don't think there's anything actionable for Sinatra to do at this point.

**Comment by dentarg** [comment]
What web server did you use? https://github.com/puma/puma/pull/3062 is included in Puma 6.1.0 and up

**Comment by dentarg** [comment, marked as answer]
In Puma 6.1.0 and up you also have https://github.com/puma/puma/pull/3040
